## Introduction to DataFrames

In [1]:
import pandas as pd

In [2]:
nba = pd.read_csv("nba.csv")

In [3]:
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [4]:
nba.values

array([['Avery Bradley', 'Boston Celtics', 0.0, ..., 180.0, 'Texas',
        7730337.0],
       ['Jae Crowder', 'Boston Celtics', 99.0, ..., 235.0, 'Marquette',
        6796117.0],
       ['John Holland', 'Boston Celtics', 30.0, ..., 205.0,
        'Boston University', nan],
       ...,
       ['Tibor Pleiss', 'Utah Jazz', 21.0, ..., 256.0, nan, 2900000.0],
       ['Jeff Withey', 'Utah Jazz', 24.0, ..., 231.0, 'Kansas', 947276.0],
       [nan, nan, nan, ..., nan, nan, nan]], dtype=object)

In [5]:
nba.shape

(458, 9)

this is a 2-D array of data set

In [6]:
nba.dtypes #to get data type info of the fields

Name         object
Team         object
Number      float64
Position     object
Age         float64
Height       object
Weight      float64
College      object
Salary      float64
dtype: object

In [7]:
nba.columns

Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
       'College', 'Salary'],
      dtype='object')

this gives you all the column(fields) names

In [8]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     373 non-null object
Salary      446 non-null float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


##### this is like the describe() method in the series object

In [12]:
rev = pd.read_csv("revenue.csv", index_col= "Date")

In [13]:
rev.head()

New York  Los Angeles  Miami
Date                                
1/1/16       985          122    499
1/2/16       738          788    534
1/3/16        14           20    933
1/4/16       730          904    885
1/5/16       114           71    253

To get the total sum of rev for each state

In [14]:
rev.sum()

New York       5475
Los Angeles    5134
Miami          5641
dtype: int64

this sums on the y-axis by default. However, if we want to sum for all three states in a row (i.e x-axis)

In [15]:
rev.sum(axis = 1)

Date
1/1/16     1606
1/2/16     2060
1/3/16      967
1/4/16     2519
1/5/16      438
1/6/16     1935
1/7/16     1234
1/8/16     2313
1/9/16     2623
1/10/16     555
dtype: int64

the axis parameter determines how it sums the values

### Extracting data from a field (or fields) in DataFrames

In [16]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [18]:
nba["Number"].head()

0     0.0
1    99.0
2    30.0
3    28.0
4     8.0
Name: Number, dtype: float64

notice: since you selected just one field, it returned a series (1-D)

In [19]:
nba[["Number", "Position", "Age"]].tail(10)

Number Position   Age
448    20.0       SF  26.0
449     5.0       SG  23.0
450     2.0       SF  28.0
451    23.0       SF  26.0
452    41.0       PF  20.0
453     8.0       PG  26.0
454    25.0       PG  24.0
455    21.0        C  26.0
456    24.0        C  26.0
457     NaN      NaN   NaN

see how we selected multiple fields

In [29]:
nba[nba["Number"] > 30][["Name","Number","Position","Age"]].head(10)

Name  Number Position   Age
1              Jae Crowder    99.0       SF  25.0
5             Amir Johnson    90.0       PF  29.0
6            Jordan Mickey    55.0       PF  21.0
7             Kelly Olynyk    41.0        C  25.0
9             Marcus Smart    36.0       PG  22.0
14            Tyler Zeller    44.0        C  26.0
15        Bojan Bogdanovic    44.0       SG  27.0
25             Willie Reed    33.0       PF  26.0
26         Thomas Robinson    41.0       PF  25.0
32  Thanasis Antetokounmpo    43.0       SF  23.0

the code above shows how to filter data from the Dataframe

In [30]:
nba.columns

Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
       'College', 'Salary'],
      dtype='object')

In [31]:
nba[nba["Number"] > 30].head()

Name            Team  Number Position   Age Height  Weight  \
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

          College      Salary  
1       Marquette   6796117.0  
5             NaN  12000000.0  
6             LSU   1170960.0  
7         Gonzaga   2165160.0  
9  Oklahoma State   3431040.0

### Add New Columns to a DataFrame

In [32]:
nba["Sport"] = "Basketball"

In [33]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary       Sport  
0              Texas  7730337.0  Basketball  
1          Marquette  6796117.0  Basketball  
2  Boston University        NaN  Basketball  
3      Georgia State  1148640.0  Basketball  
4                NaN  5000000.0  Basketball

In [34]:
nba["Age_Double"] = nba["Age"] * 2

In [35]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary       Sport  Age_Double  
0              Texas  7730337.0  Basketball        50.0  
1          Marquette  6796117.0  Basketball        50.0  
2  Boston University        NaN  Basketball        54.0  
3      Georgia State  1148640.0  Basketball        44.0  
4                NaN  5000000.0  Basketball        58.0

In [36]:
del nba["Sport"]

In [37]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  Age_Double  
0              Texas  7730337.0        50.0  
1          Marquette  6796117.0        50.0  
2  Boston University        NaN        54.0  
3      Georgia State  1148640.0        44.0  
4                NaN  5000000.0        58.0

##### You can also use the .insert() method

nba.insert(loc, column, value) 

where:<br> 
>loc = index position you want to place field<br>
>column = name of column<br>
>value = data

In [39]:
nba.insert(7, column = "Weight(Kg)",value = nba["Weight"] * 0.453592)

In [40]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   Weight(Kg)            College     Salary  Age_Double  
0   81.646560              Texas  7730337.0        50.0  
1  106.594120          Marquette  6796117.0        50.0  
2   92.986360  Boston University        NaN        54.0  
3   83.914520      Georgia State  1148640.0        44.0  
4  104.779752                NaN  5000000.0        58.0

##### Using .value_counts() method in DataFrames

In [46]:
nba["Team"].value_counts().head(7).sort_values()

Charlotte Hornets       15
Boston Celtics          15
San Antonio Spurs       15
New York Knicks         16
Milwaukee Bucks         16
Memphis Grizzlies       18
New Orleans Pelicans    19
Name: Team, dtype: int64

In [47]:
nba[nba["Team"] == "Boston Celtics"]["Team"].value_counts()

Boston Celtics    15
Name: Team, dtype: int64

#### Drop Rows with Null Values

Using the **.dropna()** method to drop rows with NaN values

In [48]:
nba.tail()

Name       Team  Number Position   Age Height  Weight  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN   

     Weight(Kg) College     Salary  Age_Double  
453   92.079176  Butler  2433333.0        52.0  
454   81.192968     NaN   900000.0        48.0  
455  116.119552     NaN  2900000.0        52.0  
456  104.779752  Kansas   947276.0        52.0  
457         NaN     NaN        NaN         NaN

In [50]:
nba.dropna().tail()

Name       Team  Number Position   Age Height  Weight  \
449    Rodney Hood  Utah Jazz     5.0       SG  23.0    6-8   206.0   
451  Chris Johnson  Utah Jazz    23.0       SF  26.0    6-6   206.0   
452     Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0   
456    Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0   

     Weight(Kg)   College     Salary  Age_Double  
449   93.439952      Duke  1348440.0        46.0  
451   93.439952    Dayton   981348.0        52.0  
452  106.140528  Kentucky  2239800.0        40.0  
453   92.079176    Butler  2433333.0        52.0  
456  104.779752    Kansas   947276.0        52.0


Notice how row 457 is dropped because it has NaN values

In [52]:
nba.dropna(how = "all").head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   Weight(Kg)            College     Salary  Age_Double  
0   81.646560              Texas  7730337.0        50.0  
1  106.594120          Marquette  6796117.0        50.0  
2   92.986360  Boston University        NaN        54.0  
3   83.914520      Georgia State  1148640.0        44.0  
4  104.779752                NaN  5000000.0        58.0

The .dropna() method removes any row with one or more NaN values by default. However, the __how__ parameter can be used to tweak it. As seen above. It removes only rows that all its fields have NaN values.<br> It also takes in the __inplace__ parameter to make changes permanent.

In [62]:
nba.dropna(axis = 1).head() #this will remove fields(or columns) with NaN values

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

To remove rows that have NaN in particular fields, we use the __subset__ param

In [70]:
nba = pd.read_csv("nba.csv")

In [77]:
nba[nba["Salary"] == 7730337.0]

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   

  College     Salary  
0   Texas  7730337.0

In [96]:
nba[nba["College"].isnull()].sort_values(by ='Name').head()

Name                   Team  Number Position   Age Height  \
330       Al Jefferson      Charlotte Hornets    25.0        C  31.0   6-10   
275      Alexis Ajinca   New Orleans Pelicans    42.0        C  28.0    7-2   
348  Amar'e Stoudemire             Miami Heat     5.0       PF  33.0   6-10   
5         Amir Johnson         Boston Celtics    90.0       PF  29.0    6-9   
90    Anderson Varejao  Golden State Warriors    18.0       PF  33.0   6-11   

     Weight College      Salary  
330   289.0     NaN  13500000.0  
275   248.0     NaN   4389607.0  
348   245.0     NaN    947276.0  
5     240.0     NaN  12000000.0  
90    273.0     NaN    289755.0

In [98]:
nba.dropna(subset= ["College","Salary"]).head(7)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

          College     Salary  
0           Texas  7730337.0  
1       Marquette  6796117.0  
3   Georgia State  1148640.0  
6             LSU  1170960.0  
7         Gonzaga  2165160.0  
8      Louisville  1824360.0  
9  Oklahoma State  3431040.0

The above command drops all rows with NaN under __College__ and __Salary__ fields

##### Fill in Null values with the __.fillna()__ method<br> If you have a consistent dataType for all fields, you can use this method in the DF as a whole.<br>If you do not, use it on individual fields

In [116]:
nba["College"].fillna(value="No college", inplace= True)            

The above code fills all the NaN values in the __College__ field with "No college"

#### Using the .astype() method to change dataTypes of fields in a DF

- This method does not act on fields that have NaN values
- First check the DF fields by using .info attribute
- Take off rows with NaN all through with .dropna() method
- Change NaN values in fields using .fillna() method
- After using the .astype() method, re-assign that field as the method just returns a copy

In [130]:
nba = pd.read_csv("nba.csv")

In [131]:
nba.info() #First check the DF fields by using .info attribute

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     373 non-null object
Salary      446 non-null float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


In [132]:
nba.dropna(how= "all", inplace= True) #Take off rows with NaN all through with .dropna() method

In [135]:
nba["College"].fillna(value="No College", inplace = True)
nba["Salary"].fillna(value= 0, inplace = True) #Change NaN values in fields using .fillna() method

In [139]:
nba["Salary"].astype("int").head()

0    7730337
1    6796117
2          0
3    1148640
4    5000000
Name: Salary, dtype: int32

Here, we change the dtype of _Salary_ field from float to int. Note, this does not affect the field in the DF.

In [144]:
nba["Salary"] = nba["Salary"].astype("int")
nba["Age"] = nba["Age"].astype("int")
nba["Number"] = nba["Number"].astype("int")

In [145]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null int32
Position    457 non-null object
Age         457 non-null int32
Height      457 non-null object
Weight      457 non-null float64
College     457 non-null object
Salary      457 non-null int32
dtypes: float64(1), int32(3), object(5)
memory usage: 30.3+ KB


In [149]:
nba["College"].value_counts().head()

No College        84
Kentucky          22
Duke              20
Kansas            18
North Carolina    16
Name: College, dtype: int64

In [150]:
nba["College"].nunique()

119

the __.nunique()__ method returns how many unique data in a field

## Practicing with Summit CS Data

In [151]:
cs = pd.read_excel("cs_test.xlsx")

In [153]:
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9193 entries, 0 to 9192
Data columns (total 14 columns):
Company number          9193 non-null int64
Customer number         9193 non-null int64
Active record code      1574 non-null object
Customer name           9193 non-null object
Customer address 1      8869 non-null object
Customer address 2      5947 non-null object
City                    9113 non-null object
State                   9157 non-null object
Zip code                9031 non-null object
Zip code2               9193 non-null int64
Creation Date           9193 non-null int64
Creation Time           9193 non-null int64
Date Last Maintained    9193 non-null int64
Time Last Maintained    9193 non-null int64
dtypes: int64(7), object(7)
memory usage: 1005.6+ KB


In [154]:
cs.columns

Index(['Company number', 'Customer number', 'Active record code',
       'Customer name', 'Customer address 1', 'Customer address 2', 'City',
       'State', 'Zip code', 'Zip code2', 'Creation Date', 'Creation Time',
       'Date Last Maintained', 'Time Last Maintained'],
      dtype='object')

In [178]:
 cs["Customer address 1"].fillna(value="No address", inplace = True)

In [179]:
cs[["Customer number","Customer name","Customer address 1", "City"]].tail()

Customer number           Customer name Customer address 1      City
9188          5001005                 POPEYES         No address       NaN
9189          5005005                 POPEYES         No address       NaN
9190         33333333      MIMIX TEST ACCOUNT         No address       NaN
9191        300300140                   PIZZA         No address   TORONTO
9192        300300800  PHILOSPHER TRAIL PIZZA         No address  BRAMPTON

In [180]:
cs.shape

(9193, 14)

In [186]:
cs["Customer name"].nunique()

6267

In [189]:
cs[(cs["Active record code"] == "A") & (cs["City"] == "LONDON")][["Customer number","Customer name","Customer address 1", "City"]].head(10)

Customer number              Customer name         Customer address 1  \
88             10785         KENSINGTON VILLAGE          1340 HURON STREET   
99             10885  LONDON EXTENDICARE (FOOD)            860 WATERLOO ST   
100            10890         LONDON EXTENDICARE            860 WATERLOO ST   
101            10900                LONDON LIFE        255 DUFFERIN AVENUE   
175            11550  HEALTHCARE MATERIALS MGMT      188 STRONACH CRESCENT   
341            30025       A CHILD CARE SERVICE                385 HILL ST   
351            30065    LONDON BRIDGE CHILDCARE  550 FANSHAWE PARK RD EAST   
354            30070    LONDON BRIDGE CHILDCARE  550 FANSHAWE PARK RD EAST   
356            30075   LONDON BRIDGE CHILD CARE  550 FANSHAWE PARK RD EAST   
358            30080    LONDON BRIDGE CHILDCARE  550 FANSHAWE PARK RD EAST   

       City  
88   LONDON  
99   LONDON  
100  LONDON  
101  LONDON  
175  LONDON  
341  LONDON  
351  LONDON  
354  LONDON  
356  LONDON  
358  LONDON

In [197]:
cs[(cs["Customer name"].str.contains("Swiss", case=False)) & (cs["Active record code"] == "A")].head()

Company number  Customer number Active record code  \
5551               1           205020                  A   
5553               1           205030                  A   
5554               1           205071                  A   
5558               1           207073                  A   
5560               1           207091                  A   

                  Customer name       Customer address 1  \
5551  SWISS CHALET            *  C/O CARA OPERATIONS LTD   
5553               SWISS CHALET               No address   
5554  SWISS CHALET            *  C/O CARA OPERATIONS LTD   
5558       SWISS CHALET EXPRESS      2148 QUEEN STREET E   
5560  SWISS CHALET            *  C/O CARA OPERATIONS LTD   

        Customer address 2         City State Zip code   Zip code2  \
5551     6303 AIRPORT ROAD  MISSISSAUGA    ON  L4V 1R8  7168374300   
5553  3010 WINTON RD SOUTH    ROCHESTER    US    14623  5854277970   
5554     6303 AIRPORT ROAD  MISSISSAUGA    ON  L4V 1R8  9547761630   
5558                   NaN      TORONTO    ON  M4E 1E3  4166932881   
5560     6303 AIRPORT ROAD  MISSISSAUGA    ON  L4V 1R8  9058460262   

      Creation Date  Creation Time  Date Last Maintained  Time Last Maintained  
5551              0              0                     0                     0  
5553              0              0              20120529                 94225  
5554              0              0                     0                     0  
5558              0              0              20081102                212312  
5560              0              0              20081102                212118

We used the __.str.contains(value, case)__ method to filter a field. It checks if a field contains a particular string.

The __category__ data type is used for fields with a lot of duplicated data. It reduces the memory usage and increase query speed.

In [198]:
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9193 entries, 0 to 9192
Data columns (total 14 columns):
Company number          9193 non-null int64
Customer number         9193 non-null int64
Active record code      1574 non-null object
Customer name           9193 non-null object
Customer address 1      9193 non-null object
Customer address 2      5947 non-null object
City                    9113 non-null object
State                   9157 non-null object
Zip code                9031 non-null object
Zip code2               9193 non-null int64
Creation Date           9193 non-null int64
Creation Time           9193 non-null int64
Date Last Maintained    9193 non-null int64
Time Last Maintained    9193 non-null int64
dtypes: int64(7), object(7)
memory usage: 1005.6+ KB


see how mem usage is 1005.6+ KB

In [199]:
cs["Active record code"].nunique()

2

In [200]:
cs["Active record code"] = cs["Active record code"].astype("category")

In [201]:
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9193 entries, 0 to 9192
Data columns (total 14 columns):
Company number          9193 non-null int64
Customer number         9193 non-null int64
Active record code      1574 non-null category
Customer name           9193 non-null object
Customer address 1      9193 non-null object
Customer address 2      5947 non-null object
City                    9113 non-null object
State                   9157 non-null object
Zip code                9031 non-null object
Zip code2               9193 non-null int64
Creation Date           9193 non-null int64
Creation Time           9193 non-null int64
Date Last Maintained    9193 non-null int64
Time Last Maintained    9193 non-null int64
dtypes: category(1), int64(7), object(6)
memory usage: 942.8+ KB


Now, the mem usage has decreased to 942.8+ KB

In [203]:
cs.nunique()

Company number             1
Customer number         9193
Active record code         2
Customer name           6267
Customer address 1      5732
Customer address 2      4270
City                     854
State                     25
Zip code                4836
Zip code2               5818
Creation Date           2075
Creation Time           5603
Date Last Maintained    1648
Time Last Maintained    7799
dtype: int64


### Using the .sort_values() method in DataFrames

In [205]:
nba.sort_values("Salary", ascending= False).head(10)

Name                   Team  Number Position  Age Height  \
109      Kobe Bryant     Los Angeles Lakers      24       SF   37    6-6   
169     LeBron James    Cleveland Cavaliers      23       SF   31    6-8   
33   Carmelo Anthony        New York Knicks       7       SF   32    6-8   
251    Dwight Howard        Houston Rockets      12        C   30   6-11   
339       Chris Bosh             Miami Heat       1       PF   32   6-11   
100       Chris Paul   Los Angeles Clippers       3       PG   31    6-0   
414     Kevin Durant  Oklahoma City Thunder      35       SF   27    6-9   
164     Derrick Rose          Chicago Bulls       1       PG   27    6-3   
349      Dwyane Wade             Miami Heat       3       SG   34    6-4   
174       Kevin Love    Cleveland Cavaliers       0       PF   27   6-10   

     Weight       College    Salary  
109   212.0    No College  25000000  
169   250.0    No College  22970500  
33    240.0      Syracuse  22875000  
251   265.0    No College  22359364  
339   235.0  Georgia Tech  22192730  
100   175.0   Wake Forest  21468695  
414   240.0         Texas  20158622  
164   190.0       Memphis  20093064  
349   220.0     Marquette  20000000  
174   251.0          UCLA  19689000

In [216]:
nba[nba["Team"].str.contains("Lakers", case= False)][["Name", "Salary"]].sort_values("Salary", ascending = False)

Name    Salary
109        Kobe Bryant  25000000
111        Roy Hibbert  15592217
118     Louis Williams   7000000
120         Nick Young   5219169
116   D'Angelo Russell   5103120
115      Julius Randle   3132240
106       Brandon Bass   3000000
113         Ryan Kelly   1724250
114    Larry Nance Jr.   1155600
117       Robert Sacre    981348
119  Metta World Peace    947276
107        Tarik Black    845059
110    Jordan Clarkson    845059
108      Anthony Brown    700000
112    Marcelo Huertas    525093

### Using the .sort_values() method on multiple DataFrames fields(or columns)

In [221]:
nba.sort_values(["Salary","Name"], ascending= [False, False]).head(10)

Name                   Team  Number Position  Age Height  \
109        Kobe Bryant     Los Angeles Lakers      24       SF   37    6-6   
169       LeBron James    Cleveland Cavaliers      23       SF   31    6-8   
33     Carmelo Anthony        New York Knicks       7       SF   32    6-8   
251      Dwight Howard        Houston Rockets      12        C   30   6-11   
339         Chris Bosh             Miami Heat       1       PF   32   6-11   
100         Chris Paul   Los Angeles Clippers       3       PG   31    6-0   
414       Kevin Durant  Oklahoma City Thunder      35       SF   27    6-9   
164       Derrick Rose          Chicago Bulls       1       PG   27    6-3   
349        Dwyane Wade             Miami Heat       3       SG   34    6-4   
294  LaMarcus Aldridge      San Antonio Spurs      12       PF   30   6-11   

     Weight       College    Salary  
109   212.0    No College  25000000  
169   250.0    No College  22970500  
33    240.0      Syracuse  22875000  
251   265.0    No College  22359364  
339   235.0  Georgia Tech  22192730  
100   175.0   Wake Forest  21468695  
414   240.0         Texas  20158622  
164   190.0       Memphis  20093064  
349   220.0     Marquette  20000000  
294   240.0         Texas  19689000

when sorting multiple fields:
- the method first sorts by the first element in the list and so on
- you can also change the ascending order of the data by also putting the booleans in a list

### Rank Values with the .rank() Method

In [222]:
nba = pd.read_csv("nba.csv")

In [226]:
nba.dropna(how = "all", inplace= True)
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     373 non-null object
Salary      446 non-null float64
dtypes: float64(4), object(5)
memory usage: 35.7+ KB


In [228]:
nba["Salary"].fillna(0, inplace= True)
nba["College"].fillna("No college", inplace = True)

In [232]:
nba["Salary"].astype("int", inplace= True).head()

0    7730337
1    6796117
2          0
3    1148640
4    5000000
Name: Salary, dtype: int32

In [235]:
nba["Salary Rank"] = nba["Salary"].rank(ascending = False).astype("int")

In [236]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  Salary Rank  
0              Texas  7730337.0           97  
1          Marquette  6796117.0          110  
2  Boston University        0.0          452  
3      Georgia State  1148640.0          322  
4         No college  5000000.0          147

In [243]:
nba[nba["Name"].str.contains("Stephen Curry", case = False)][["Name","Salary","Salary Rank"]]

Name      Salary  Salary Rank
80  Stephen Curry  11370786.0           61